src/DecryptionFailureTracker.ts

```diff
@@ -176,21 +220,90 @@ export class DecryptionFailureTracker {
+        if (failure) {
+            this.failures.delete(eventId);
+
+            const timeToDecryptMillis = nowTs - failure.ts;
+            if (timeToDecryptMillis < DecryptionFailureTracker.GRACE_PERIOD_MS) {
+                // the event decrypted on time, so we don't need to report it
+                return;
+            } else if (timeToDecryptMillis <= DecryptionFailureTracker.MAXIMUM_LATE_DECRYPTION_PERIOD) {
+                // The event is a late decryption, so store the time it took.
+                // If the time to decrypt is longer than
+                // MAXIMUM_LATE_DECRYPTION_PERIOD, we consider the event as
+                // undecryptable, and leave timeToDecryptMillis undefined
+                failure.timeToDecryptMillis = timeToDecryptMillis;
+            }
+            this.failuresToReport.add(failure);
+        }
+    }
+
+    private async handleKeysChanged(client: MatrixClient): Promise<void> {
+        if (this.checkingVerificationStatus) {
+            this.retryVerificationStatus = true;
+        } else {
+            this.checkingVerificationStatus = true;
+            try {
+                do {
+                    this.retryVerificationStatus = false;
+                    this.userTrustsOwnIdentity = (
+                        await client.getCrypto()!.getUserVerificationStatus(client.getUserId()!)
+                    ).isCrossSigningVerified();
+                } while (this.retryVerificationStatus);
+            } finally {
+                this.checkingVerificationStatus = false;
+            }
+        }
+    }
+
+    private async setClient(client: MatrixClient): Promise<void> {
+        if (this.client && this.keysChangedHandler) {
+            this.client.removeListener(CryptoEvent.KeysChanged, this.keysChangedHandler);
+        }
+        this.keysChangedHandler = undefined;
+        this.client = client;
+
+        const baseProperties: ErrorProperties = {};
+
+        this.userDomain = client.getDomain() ?? undefined;
+        if (this.userDomain === "matrix.org") {
+            baseProperties.isMatrixDotOrg = true;
+        } else if (this.userDomain !== undefined) {
+            baseProperties.isMatrixDotOrg = false;
+        }
+
+        const crypto = client.getCrypto();
+        if (crypto) {
+            const version = crypto.getVersion();
+            if (version.startsWith("Rust SDK")) {
+                baseProperties.cryptoSDK = "Rust";
+            } else {
+                baseProperties.cryptoSDK = "Legacy";
+            }
+            this.userTrustsOwnIdentity = (
+                await crypto.getUserVerificationStatus(client.getUserId()!)
+            ).isCrossSigningVerified();
+            this.keysChangedHandler = () => {
+                this.handleKeysChanged(client).catch((e) => {
+                    console.log("Error handling KeysChanged event", e);
+                });
+            };
+            client.on(CryptoEvent.KeysChanged, this.keysChangedHandler);
```

Comment:
if we're going to go down the path of adding our own event handlers, then we should be consistent and add our own `EventDecrypted` handler, instead of requiring the caller to add one which calls `eventDecypted`.

But actually, it would probably save us a whole lot of effort just to expose `handleKeysChanged` and have the caller add the event handler? (In particular, it means we don't need to worry about *removing* the listener, because the only time we have to do so is when the client is being deleted anyway).

Reply:
I've opted to go with registering our own `EventDecrypted` handler, and `SessionLoggedOut` handler, so that the only thing that the application needs to do here is call `.start`. And since I'm registering our own `SessionLoggedOut` handler, I can pass the client and other handlers in a closure.  (We probably don't actually need to remove the listeners, but I'd rather clean things up properly, rather than assuming that it'll be fine.)